### ISSUE TITLE
Catalog job only get repo top 100

### ISSUE THREAD

**Issue by summerQLin** [SUSE/Portus#548]
Catalog job `bundle exec rails r bin/client.rb catalog` will only get first 100 repos. Therefore if user pushes an image but lists after 100, he will be able to see new pushed image, thanks to notification events. But that image will soon removed by crono catalog job, since it only get 100 repos.

**Comment by mssola** [comment]
@summerQLin wow, that's weird. Let me investigate on this.

**Comment by summerQLin** [comment]
You can check registry catalog api. The api call will return 100 repos by default, then get next n repos by sending parameter `n` and `last`. https://docs.docker.com/registry/spec/api/#catalog

**Comment by Salzig** [comment]
The Documentation states:
>
Catalog Fetch Complete
GET /v2/_catalog
Request an unabridged list of repositories available.

so, it should only limit the response, when you request it to do.

**Comment by mssola** [comment]
@Salzig yes, that's correct. That's also what I understood from the documentation. However, I want to check this out by myself before closing to make sure whether this bug exists, and if it does, whether it's distribution's fault, or our own.

**Comment by mssola** [comment]
@summerQLin in order to investigate this, I've added a repository with 110 tags, and I've pushed 110 different images in my local registry. This is the response (after adding a print statement on the `catalog` command): http://pastebin.nue.suse.com/16735. As you can see, everything worked properly really.

So, in order to further debug this, I need the following:

- What does your log file say about this ? Could you paste the logging message generated when the catalog job kicks in ? This is important, because as you can see, in `app/jobs/catalog_job.rb` line 16 we actually log the exception.
- Could please provide the configuration of your registry?

**Comment by summerQLin** [comment]
@mssola, for your first question, log didn't throw any exception, catalog job is able to update database.
For my registry configuration, I set environment variables as below.
```
    REGISTRY_AUTH_TOKEN_ISSUER: 
    REGISTRY_AUTH_TOKEN_REALM: 
    REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE: 
    REGISTRY_AUTH_TOKEN_SERVICE: 
    REGISTRY_HTTP_SECRET: 
    REGISTRY_STORAGE_SWIFT_AUTHURL: 
    REGISTRY_STORAGE_SWIFT_CONTAINER: 
    REGISTRY_STORAGE_SWIFT_INSECURESKIPVERIFY: 'true'
    REGISTRY_STORAGE_SWIFT_PASSWORD: 
    REGISTRY_STORAGE_SWIFT_TENANTID: 
    REGISTRY_STORAGE_SWIFT_USERNAME: 
    REGISTRY_STORAGE_DELETE_ENABLED: true
    REGISTRY_REDIS_ADDR: cache:6379
    SEARCH_BACKEND: sqlalchemy
    REGISTRY_HTTP_TLS_CERTIFICATE: 
    REGISTRY_HTTP_TLS_KEY: 
```
By the way, you link http://pastebin.nue.suse.com/16735 can't be open.
And you can see code here, https://github.com/docker/distribution/blob/master/registry/handlers/catalog.go, seems there is a max value.

**Comment by mssola** [comment]
> By the way, you link http://pastebin.nue.suse.com/16735 can't be open.
My bad :_) This should work: http://pastebin.com/hesCYwxY

> And you can see code here, https://github.com/docker/distribution/blob/master/registry/handlers/catalog.go, seems there is a max value.

And as you can see in the [GetCatalog](https://github.com/docker/distribution/blob/master/registry/handlers/catalog.go#L42) function, it's only used if "n" is specified. Since you're using swift, I wonder if this is a bug on this particular storage backend :/

To help us with this, is there something on the registry log about this ?

**Comment by mssola** [comment]
@summerQLin I just made a quick test inside of the registry, and yes, this is a bug on the registry. Let me investigate a bit more on this, and I'll submit the issue/pr myself ;)

**Comment by mssola** [comment]
@summerQLin I've submitted an issue on distribution, as you can see. I'll continue to investigate how to solve this upstream, but anyways we are forced now to provide a workaround for this in Portus. That's why I'm adding this issue as a bug, and we'll hopefully get it fixed asap. Thanks for reporting.

**Comment by mssola** [comment]
I've added a workaround in the PR #575

**Comment by summerQLin** [comment]
Thanks :+1:

**Comment by mssola** [comment]
Fixed with PR #575.
